subprojects/ide-native/src/integTest/groovy/org/gradle/ide/xcode/XcodeCppExternalSourceDependenciesIntegrationTest.groovy

```diff
@@ -0,0 +1,141 @@
+            fixture.greeter.writeToProject(it)
+        }
+        def commit = repo.commit('initial commit', GFileUtils.listFiles(file('greeter'), null, true))
+
+        and:
+        settingsFile << """
+            sourceControl {
+                vcsMappings {
+                    withModule("org.test:greeter") {
+                        from vcs(GitVersionControlSpec) {
+                            url = "${repo.url}"
+                        }
+                    }
+                }
+            }
+        """
+        buildFile << """
+            apply plugin: 'cpp-application'
+            apply plugin: 'xcode'
+            group = 'org.gradle'
+            version = '2.0'
+
+            dependencies {
+                implementation "org.test:greeter:latest.integration"
+            }
+        """
+        fixture.main.writeToProject(testDirectory)
+
+        when:
+        succeeds 'xcode'
+
+        then:
+        executedAndNotSkipped(":xcodeProject", ":xcodeProjectWorkspaceSettings", ":xcodeSchemeAppExecutable", ":xcode")
+        rootXcodeWorkspace.contentFile.assertHasProjects("${rootProjectName}.xcodeproj")
+
+        def appProject = xcodeProject("${rootProjectName}.xcodeproj").projectFile
+        appProject.indexTarget.getBuildSettings().HEADER_SEARCH_PATHS == toSpaceSeparatedList(file('src/main/headers'), checkoutDir(repo.name, commit.id.name, repo.id).file('src/main/public'))
+    }
+
+    def "adds source dependencies Xcode project of main component to Xcode workspace"() {
+        def fixture = new CppAppWithLibrary()
+
+        given:
+        buildTestFixture.withBuildInSubDir()
+        depProject = singleProjectBuild("greeter") {
+            buildFile << """
+                apply plugin: 'cpp-library'
+                apply plugin: 'xcode'
+            """
+            fixture.greeter.writeToProject(it)
+        }
+        def commit = repo.commit('initial commit', GFileUtils.listFiles(file('greeter'), null, true))
+
+        and:
+        settingsFile << """
+            sourceControl {
+                vcsMappings {
+                    withModule("org.test:greeter") {
+                        from vcs(GitVersionControlSpec) {
+                            url = "${repo.url}"
+                        }
+                    }
+                }
+            }
+        """
+        buildFile << """
+            apply plugin: 'cpp-application'
+            apply plugin: 'xcode'
+            group = 'org.gradle'
+            version = '2.0'
+
+            dependencies {
+                implementation "org.test:greeter:latest.integration"
+            }
+        """
+        fixture.main.writeToProject(testDirectory)
+
+        when:
+        succeeds 'xcode'
+
+        then:
+        executedAndNotSkipped(":xcodeProject", ":xcodeProjectWorkspaceSettings", ":xcodeSchemeAppExecutable", ":xcode",
+            ":greeter:xcodeProject", ":greeter:xcodeProjectWorkspaceSettings", ":greeter:xcodeSchemeGreeterSharedLibrary")
```

Comment:
It should build the binaries of the `greeter` library rather than the project files. We don't want to include the source dependency as an Xcode project but instead as a pre-built library.

Reply:
This is exactly the issue I was having and trying to figure out. The code that finds all `xcodeproj` will search any project/included build that defines `xcodeproj` artifact. If the source dependency applies the `xcode` plugin, the root build will find the `xcodeproj` files and will try to include them. I was asking how to find out is a particular included build was included because of source dependency so we can exclude them. Otherwise, they are going to be included and there isn't much we can do about it (with the current code).

If we only populate the import paths and search directories, all works fine in term of symbol discovery, but the Swift experience may not be the best as you won't be able to jump to the source code. Instead, you will jump to the Swift file representation of the `.swiftmodule` which only show the class/function prototype.

The middle ground is to include them as normal, but any change to the project source file will be overwritten on each build as Gradle will do a `git reset --hard` on the folder. This way you can jump to the implementation, but won't be able to change it.

Reply:
I would solve the issue of navigating to the source separately (and later). We should try hard to present the source dependencies as _libraries_ in xcode, the same way as binary dependencies are presented, and not as _projects_ in xcode. We have to fix navigation to the source for binary dependencies as well, I would fix it for source dependencies in the same way and at the same time.

Reply:
As far as skipping the project file for source dependencies goes, we should filter the set of project file artifacts in the same way that we filter the binaries (just with the filter inverted of course).